mava/systems/sac/ff_isac.py

```diff
@@ -0,0 +1,600 @@
+
+        return params, opt_states, q_loss_info
+
+    def update_actor_and_alpha(
+        params: SacParams, opt_states: OptStates, data: Transition, key: chex.PRNGKey
+    ) -> Tuple[SacParams, OptStates, Metrics]:
+        """Update the actor and alpha parameters. Compensated for the delay in policy updates."""
+        # compensate for the delay by doing `policy_frequency` updates instead of 1.
+        assert cfg.system.policy_frequency > 0, "Need to have a policy frequency > 0."
+        for _ in range(cfg.system.policy_frequency):
+            actor_key, alpha_key = jax.random.split(key)
+
+            # Update actor.
+            actor_grad_fn = jax.value_and_grad(actor_loss_fn)
+            actor_loss, act_grads = actor_grad_fn(
+                params.actor, data.obs, jnp.exp(params.log_alpha), params.q.online, actor_key
+            )
+            # Mean over the device and batch dimensions.
+            actor_loss, act_grads = lax.pmean((actor_loss, act_grads), axis_name="device")
+            actor_loss, act_grads = lax.pmean((actor_loss, act_grads), axis_name="batch")
+            actor_updates, new_actor_opt_state = actor_opt.update(act_grads, opt_states.actor)
+            new_actor_params = optax.apply_updates(params.actor, actor_updates)
+
+            params = params._replace(actor=new_actor_params)
+            opt_states = opt_states._replace(actor=new_actor_opt_state)
+
+            # Update alpha if autotuning
+            alpha_loss = 0.0  # loss is 0 if autotune is off
+            if cfg.system.autotune:
+                # Get log prob for alpha loss
+                pi = actor.apply(params.actor, data.obs)
+                action = pi.sample(seed=key)
+                log_prob = pi.log_prob(action)
+
+                alpha_grad_fn = jax.value_and_grad(alpha_loss_fn)
+                alpha_loss, alpha_grads = alpha_grad_fn(params.log_alpha, log_prob, target_entropy)
+                alpha_loss, alpha_grads = lax.pmean((alpha_loss, alpha_grads), axis_name="device")
+                alpha_loss, alpha_grads = lax.pmean((alpha_loss, alpha_grads), axis_name="batch")
+                alpha_updates, new_alpha_opt_state = alpha_opt.update(alpha_grads, opt_states.alpha)
+                new_log_alpha = optax.apply_updates(params.log_alpha, alpha_updates)
+
+                params = params._replace(log_alpha=new_log_alpha)
+                opt_states = opt_states._replace(alpha=new_alpha_opt_state)
+
+        loss_info = {"actor_loss": actor_loss, "alpha_loss": alpha_loss}
+        return params, opt_states, loss_info
+
+    # Act/learn loops:
+    def update_epoch(
+        carry: Tuple[BufferState, SacParams, OptStates, int, chex.PRNGKey], _: Any
+    ) -> Tuple[Tuple[BufferState, SacParams, OptStates, int, chex.PRNGKey], Metrics]:
+        """Update the Q function and optionally policy/alpha with TD3 delayed update."""
+        buffer_state, params, opt_states, t, key = carry
```

Comment:
Had a similar chat with @OmaymaMahjoub, if we do timestep here it could be confusing as one might think it's the object, I think `t` is more clear that it's a counter.

Reply:
Makes total sense.